keps/sig-api-machinery/2309-idl/syntax.md

```diff
@@ -0,0 +1,635 @@
+    // Other markers might include overrides for legacy Go type names,
+    // marking fields as being behind feature gates for automated linting
+    // and better documentation, etc.
+
+    struct SecretProjection {
+        // inline fields have their fields embedded in their parent object
+        // in systems where this is supported (e.g. JSON, Go).  In systems
+        // where not supported, generally just use the type name as the field
+        // name.  Only structs and unions may be inlined.
+
+        // "raw identifiers" are surrounded with backticks, and take two purposes
+        // the first is to allow fields and keys that have the same name as keywords
+        // (like below).
+        //
+        // The second is as an emergency "break glass" to violate the "field name must
+        // have an lowercase first letter" constraint.
+        @reference(group: "v1", version: "core", `kind`: "Secret")
+        _inline: LocalObjectReference,
+    }
+
+    @reference(same-namespace: true)
+    struct LocalObjectReference {
+        // fields are referenced using `.name`
+        // (this avoids parsing ambiguities with primitive type names)
+
+        @reference-part(part: .name)
+        name: string,
+    }
+
+}
+```
+
+```
+import (
+    types (
+       // group-versions are imported explicitly from KDL or CKDL files
+       {core/v1} from "k8s.io/api/core/v1";
+    )
+    markers (
+        // marker definitions are imported from special KDL or CKDL files
+        // and given a prefix whereever they're used.
+        kgo from "kubernetes.mkdl";
+    )
+)
+
+group-version(group: "examples", version: "v1") {
+    kind CSIDriver {
+        spec: Spec,
+
+        struct Spec {
+            // ...
+
+            // in a set, each item may appear only once.
+            volumeLifecycleModes: optional set(value: VolumeLifecycleMode),
+        }
+    }
+
+    kind StorageClass {
+        // ...
+
+        // types from other group-versions may be referenced with
+        // `group/version::path`.
+        reclaimPolicy: optional(default: Delete) core/v1::PersistentVolumeSpec::ReclaimPolicy,
+
+        @kgo::feature-gated(name: "VolumeScheduling")
+        volumeBindingMode: optional VolumeBindingMode,
+    }
+}
+```
+
+## Design Notes
+
+### Uhhh, where are the proto tags?
+
+Earlier versions of this proposal included a specific syntax for adding
+proto tags (`myField[42]: optional string` introduces a field named
+`myField` with a tag of `42`). However, this syntax was not entirely
+obvious.
+
+The current plan is to place proto tags in a look-aside file that will
+only ever be appended to, and will be managed by the compiler.  This has
+the *slight* downside of requiring extra scrutiny on field renames, but
+has the upside that users *don't have to care* or learn about how proto
+tags work, why we'd have to leave room for reserved fields in kinds, etc.
+
+The tags will still be present in the CKDL file, meaning that tooling has
```

Comment:
This seems to give proto some "better than equal" standing in the ecosystem?

Reply:
oh, you mean because they're natively supported in CKDL? vs having them appear under the `attributes` field (where the markers go) in CKDL?

I'm curious to hear your thoughts on that part in more detail, but my logic was kinda "since proto is actually an important detail to how k8s works, like JSON, it gets first-class representation".  It's not quite *as* important as JSON (in that the system will function without it), but it's important to all the k/k types, and I expect it'll eventually be something to tackle for CRDs. :woman_shrugging:

Reply:
How does JSON get first-class representation, above and beyond being a simple structured notation?  But yes, that's what I meant.

Reply:
I meant stuff like field names & enum values are phrased in terms of their JSON representation, with convention + generator-specific overrides to handle conversion to Go.